numpy/array_api/_statistical_functions.py

```diff
@@ -24,6 +35,8 @@ def mean(
     axis: Optional[Union[int, Tuple[int, ...]]] = None,
     keepdims: bool = False,
 ) -> Array:
+    if x.dtype not in _floating_dtypes:
+        raise TypeError("Only floating-point dtypes are allowed in mean")
     return Array._new(np.mean(x._array, axis=axis, keepdims=keepdims))
```

Comment:
These three were also discussed at https://github.com/data-apis/array-api/issues/202. I'm not following why these functions should upcast, however. Unlike sum and prod, the mean of a set of numbers does not exceed its maximum, so it cannot overflow. The internal calculation might overflow if implemented in a certain way, but the spec says nothing about how these functions should be calculated internally. I may be missing something though, so if I am please raise this on the array-apis repo.

Reply:
I have scanned that discussion, and I think the meaty stuff was likely discussed at meetings? I don't think there is a lot about the choice for *float* reductions (e.g. bfloat16/float32)?  For one it includes this comment: https://github.com/data-apis/array-api/issues/202#issuecomment-863402982  which is *not* what is implemented here (or in the current version of the standard?).
I am happy to explain why I am curious about the choice (see below), but since I don't have the full picture (and can't make it easily), I will just have to be content with you handling it...

I probably misunderstood [this comment](https://github.com/data-apis/array-api/pull/238#issuecomment-922135271) and you are actually worried about summing many integer valued floats and getting an *exact* integral result!?  Rather than overflow/catastrophic annihilation during the summation?

To me it seemed that the main argument would be *precision*, i.e. summing too many `float32`, and especially `bfloat16` naively quickly runs into catastrophic annihilation and bogus results.  But `mean`, `var`, `std` will all run into the problem just the way `sum` does:
```
arr_float32 = np.ones(10**9).astype(dtype=np.float32).reshape(-1, 2)
arr_float32.sum(axis=0, dtype=np.float64)  / arr_float32.shape[0]
# array([1., 1.])   -- dtype=... since array API would do that implicitly
arr_float32.mean()
# array([0.03355443, 0.03355443], dtype=float32)
```
(NumPy uses a more stable summation without the reshape, which makes showing this a bit harder...)

Reply:
The specification only specifies the dtype of the _return value_. What dtype is used for _intermediate results in the calculation_ is unspecified, and completely up to the library. As is the algorithm used. So it'd be perfectly valid to upcast to `int64` / `float64`, do the calculations, then downcast (and this is a common pattern for low-precision dtypes).

`sum` and `prod` can have outputs that overflow (mathematically), while `mean`, `min`, `max` etc. cannot. Furthermore this regularly happens in practice, for example scikit-image has code to take the sum over all axis for images, and those are typically ~1000x1000 px. Hence the different treatment of `sum` and `prod`.

Reply:
This code is also no longer present in the PR diff, so I'll resolve the comment.

Reply:
Now that I'm touching this comment, I see why it's confusing - I'll edit it to make clear it's about the returned array.